org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/adapters/ProgressTokenAdapter.java

```diff
@@ -0,0 +1,87 @@
+/******************************************************************************
+ * Copyright (c) 2018 TypeFox and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0,
+ * or the Eclipse Distribution License v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ * 
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ ******************************************************************************/
+package org.eclipse.lsp4j.adapters;
+
+import java.io.IOException;
+
+import org.eclipse.lsp4j.ProgressToken;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonParseException;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * A type adapter for class which implements WorkDoneProgressNotification interface.
+ */
+public class ProgressTokenAdapter extends TypeAdapter<ProgressToken> {
```

Comment:
Is this adapter really necessary? We already have a generic type adapter for `Either`, which should work for `ProgressToken` without further configuration (Gson considers the superclass of `ProgressToken` while looking for a type adapter). Please try whether parsing / serializing works without this adapter and the corresponding `@JsonAdapter` annotations.

Reply:
If you want it, it's possible but this way i must put ProgressToken class in json project because the EitherTypeAdapter is in json project. See my last separated commit 
https://github.com/eclipse/lsp4j/pull/442/commits/8f634798a5308be1500f1b0da1328def5caa7ad5

Or maybe you have another idea ?

Reply:
I see. I think the following should work:

 * Add the following code at the start at `createLeft` and `createRight` of EitherTypeAdapter (pass `null, obj` instead of `obj, null` in `createRight`):
```java
try {
	Constructor<Either<L, R>> constructor = (Constructor<Either<L, R>>) typeToken.getRawType().getConstructor(Object.class, Object.class);
	return constructor.newInstance(obj, null);
} catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException exc) {
	// Proceed with the code below
}
```
 * Make the constructor of `ProgressToken` public.

This way we avoid the dependency of EitherTypeAdapter to ProgressToken and you can leave ProgressToken in the LSP-specific bundle.

Reply:
et voila !!!